### PR TITLE
feat: Added credential rotation resources for release artifacts

### PR DESCRIPTION
**Pull request by royjit**
*Description of changes:* Credential rotation script in lambda for the release CircleCI workflow.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by palpatim — COMMENTED** [review]
Partially reviewed, posting comments I've got thus far

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_construct.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+        self.credential_rotator = aws_lambda_python.PythonFunction(
+            self,
+            "credential_rotation_lambda",
+            entry="cdk/credential_rotation/lambda_functions/",
+            index="credential_rotator.py",
+            runtime=aws_lambda.Runtime.PYTHON_3_6,
+            role=iam_construct.lambda_role,
+            timeout=core.Duration.minutes(5),
+            description="Credential rotation script for the CircleCI pipeline",
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_construct.py:25]: Updated description

File: src/release_artifacts_resources/ios/cdk/app.py [hunk 2]
```diff
@@ -5,7 +5,18 @@
 from cdk.distribution_stack import DistributionStack
 
 app = core.App()
+
 distribution_stack = DistributionStack(app, "DistributionStack")
-credential_rotation_stack = CredentialRotationStack(app, "CredentialRotationStack")
+
+bucket_arn = distribution_stack.s3.bucket.bucket_arn
+distribution_id = distribution_stack.cloudfront.cloudfront.distribution_id
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/app.py:12]: Yes, we can access resources from a different stack if they are in the same account - https://docs.aws.amazon.com/cdk/latest/guide/resources.html#resource_stack

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_construct.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        **kwargs
+    ) -> None:
+        super().__init__(scope, construct_id, **kwargs)
+
+        self.credential_rotator = aws_lambda_python.PythonFunction(
+            self,
+            "credential_rotation_lambda",
+            entry="cdk/credential_rotation/lambda_functions/",
+            index="credential_rotator.py",
+            runtime=aws_lambda.Runtime.PYTHON_3_6,
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_construct.py]: Yes, updated

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/credential_rotator.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+# configured to meaningful values, since key creation takes some time to propagate, and any attempt
+# to assume role using the newly created key may fail. Not to mention the usual vagaries of internet
+# communication that could benefit from a retry.
+MAX_RETRY_ATTEMPTS = 10
+MAX_RETRY_WAIT = 10
+
+
+# Make sure this is at least 2 hours longer than the key rotation schedule set up in the CFn.
+# Currently, "2 hours" is an arbitrary value based on the fact that a few of our integration tests
+# take ~1 hour to run, so we want to leave sufficient time for executing the tests in cases where a
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/lambda_functions/credential_rotator.py:26]: I think we can start with 2 hour and later modify this if this is too low? Currently building xcframeworks takes around 1 hour. The very next step is to upload to s3 and cloud front invalidation, so even I think 4hr for the session is pretty enough for the release.

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        self.circleci_release_role = aws_iam.Role(
+            self,
+            "circleci_release_role",
+            assumed_by=self.circleci_user,
+            role_name="CircleCIReleaseProcessRole",
+            max_session_duration=core.Duration.hours(4),
+        )
+
+        s3_list_policy = aws_iam.PolicyStatement(
+            effect=aws_iam.Effect.ALLOW, actions=["s3:ListAllMyBuckets"], resources=["*"]
```
Comment (royjit) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:32]: Should we remove `*` ?

Comment (palpatim) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:32]: No, `ListAllMyBuckets` doesn't take a resource

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        )
+        self.circleci_release_role.add_to_policy(s3_list_policy)
+
+        bucket_policy = aws_iam.PolicyStatement(
+            effect=aws_iam.Effect.ALLOW,
+            actions=["s3:GetObject", "s3:PutObject"],
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:38]:
```suggestion
            actions=["s3:PutObject"],
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            effect=aws_iam.Effect.ALLOW,
+            actions=["cloudfront:CreateInvalidation"],
+            resources=[cloudfront_arn],
+        )
+        self.circleci_release_role.add_to_policy(cloudfront_policy)
+
+    def lambda_execution_role(self) -> None:
+        self.lambda_role = aws_iam.Role(
+            self,
+            "lambda_execution_role",
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:53]:
```suggestion
            "lambda_key_rotation_execution_role",
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            aws_iam.ManagedPolicy.from_aws_managed_policy_name(
+                "service-role/AWSLambdaBasicExecutionRole"
+            )
+        )
+
+        logs_policy = aws_iam.PolicyStatement(
+            effect=aws_iam.Effect.ALLOW, actions=["cloudwatch:*", "logs:*"], resources=["*"]
+        )
+        self.lambda_role.add_to_policy(logs_policy)
+
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:67]:
```suggestion
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import cdk.credential_rotation.utils.secretmanager_constants as constants
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py:1]: `secretsmanager` not `secretmanager`

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+from aws_cdk import aws_iam, aws_secretsmanager, core
+from cdk.credential_rotation.iam_construct import IAMConstruct
+
+
+class SecretManagerConstruct(core.Construct):
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py:6]:
```suggestion
class SecretsManagerConstruct(core.Construct):
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        lambda_role_rotate_keys_policy = aws_iam.PolicyStatement(
+            effect=aws_iam.Effect.ALLOW,
+            actions=["iam:CreateAccessKey", "iam:DeleteAccessKey"],
+            resources=[self.circleci_user.user_arn],
+        )
+        self.lambda_role.add_to_policy(lambda_role_rotate_keys_policy)
+
+    def lambda_role_add_to_policy(self, policy: aws_iam.PolicyStatement) -> None:
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:75]:
```suggestion
    def add_policy_to_lambda_role(self, policy: aws_iam.PolicyStatement) -> None:
```

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/lambda_constants.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+CIRCLECI_CONFIG_SECRET_ENV = "CIRCLECI_CONFIG_SECRET"
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/lambda_constants.py:1]: Add a comment enumerating the places these are hardcoded (e.g., the Lambda rotator function)

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/secretmanager_constants.py [hunk 13]
```diff
@@ -0,0 +1 @@
+CIRCLECI_API_KEY = "CIRCLECI_API_KEY"
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/secretmanager_constants.py:1]: Add a comment enumerating the places these are hardcoded (e.g., the Lambda rotator function)

File: src/release_artifacts_resources/ios/cdk/cdk/distribution/cloudfront_construct.py [hunk 14]
```diff
@@ -9,7 +9,7 @@ def __init__(
 
         super().__init__(scope, construct_id, **kwargs)
 
-        cloudfront = aws_cloudfront.Distribution(
+        self.cloudfront = aws_cloudfront.Distribution(
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/distribution/cloudfront_construct.py:12]:
```suggestion
        self.distribution = aws_cloudfront.Distribution(
```

File: src/release_artifacts_resources/ios/cdk/app.py (hunk 2 above)
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/app.py:12]:
```suggestion
distribution_id = distribution_stack.cloudfront.distribution.distribution_id
```

File: src/release_artifacts_resources/ios/cdk/README.md [hunk 15]
```diff
@@ -45,4 +64,23 @@ command.
  * `cdk diff`        compare deployed stack with current state
  * `cdk docs`        open CDK documentation
 
-Enjoy!
+## Stack information
+
+### Distribution stack
+
+Release resources for distribution are declared in the stack named - `DistributionStack`.
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/README.md]: Added

File: src/release_artifacts_resources/ios/cdk/app.py (hunk 2 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/app.py]: fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        scope: core.Construct,
+        construct_id: str,
+        bucket_arn: str,
+        cloudfront_arn: str,
+        **kwargs
+    ) -> None:
+
+        super().__init__(scope, construct_id, **kwargs)
+        self.circleci_release_user()
+        self.circleci_release_proceess_role(bucket_arn=bucket_arn, cloudfront_arn=cloudfront_arn)
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        self.lambda_execution_role()
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+    def circleci_release_user(self) -> None:
+        self.circleci_user = aws_iam.User(self, "circleci_iam_user", user_name="CircleCIIAMUser")
```
[the remark this answers is not in the document]
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py (hunk 6 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: removed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py (hunk 7 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py (hunk 8 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:66]: Without this the lambda was not executing. So added back.

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py (hunk 11 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py (hunk 9 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py (hunk 10 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/secretmanager_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/distribution/cloudfront_construct.py (hunk 14 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/distribution/cloudfront_construct.py]: Fixed

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py [hunk 19]
```diff
@@ -61,8 +56,8 @@ def lambda_execution_role(self) -> None:
         )
 
         logs_policy = aws_iam.PolicyStatement(
-            effect=aws_iam.Effect.ALLOW, actions=["cloudwatch:*", "logs:*"], resources=["*"]
-        )
+             effect=aws_iam.Effect.ALLOW, actions=["cloudwatch:*", "logs:*"], resources=["*"]
+         )
```
Comment (palpatim) [inline src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/iam_construct.py:60]: nit: this might be incorrectly indented?

**Review by palpatim — APPROVED** [review]
minor spacing nit, otherwise LGTM!

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/lambda_constants.py (hunk 12 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/lambda_constants.py:3]: Added

File: src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/secretmanager_constants.py (hunk 13 above)
Comment (royjit) [reply src/release_artifacts_resources/ios/cdk/cdk/credential_rotation/utils/secretmanager_constants.py:2]: Added

**Pull request merged**